lib/posix-socket/socket_vnops.c

```
@@ -0,0 +1,323 @@
+	return ret;
+}
+
+int
+posix_socket_vfscore_close(struct vnode *s_vnode,
+		struct vfscore_file *vfscore_file)
+{
+	int ret = 0;
+	struct posix_socket_file *file = NULL;
+	file = s_vnode->v_data;
+
+	uk_pr_debug("%s fd:%d driver:%s: sock_data:%p\n",
+			__func__,
+			file->vfs_file->fd,
+			file->driver->libname,
+			file->sock_data);
+
+	/* Close and release the socket */
+	ret = posix_socket_close(file);
+
+	/*
+	 * Free socket file
+	 * The rest of the resources will be freed by vfs
+	 *
+	 * TODO: vfs ignores close errors right now, so free our file
+	 */
+	uk_free(file->driver->allocator, file);
+
+	if (ret < 0)
+		return errno;
+
+	return ret;
+}
+
+int
+posix_socket_vfscore_write(struct vnode *s_vnode,
+		struct uio *buf, int ioflag __unused)
+{
+	int ret = 0;
+	struct posix_socket_file *file = NULL;
+
+	file = s_vnode->v_data;
+	uk_pr_debug("%s fd:%d driver:%s sock_data:%p\n",
+		__func__,
+		file->vfs_file->fd,
+		file->driver->libname,
+		file->sock_data);
+
+	/* Write to the socket */
+	ret = posix_socket_write(file, buf->uio_iov->iov_base, buf->uio_iov->iov_len);
+
+	/*
+	 * Some socket implementations, such as LwIP, may set the errno and return to
+	 * -1 as an error, but vfs expects us to return a positive errno.
+	 */
+	if (ret < 0)
+		return errno;
+
+	buf->uio_resid -= ret;
+	return 0;
+}
+
+int
+posix_socket_vfscore_read(struct vnode *s_vnode,
+		struct vfscore_file *vfscore_file __unused,
+		struct uio *buf, int ioflag __unused)
+{
+	int ret = 0;
+	struct posix_socket_file *file = NULL;
+
+	file = s_vnode->v_data;
+	if (PTRISERR(file)) {
+		ret = -1;
+		SOCKET_ERR(PTR2ERR(file), "failed to identify socket descriptor");
+		goto EXIT;
+	}
+
+	uk_pr_debug("%s fd:%d driver:%s sock_data:%p\n",
+			__func__,
+			file->vfs_file->fd,
+			file->driver->libname,
+			file->sock_data);
+
+	ret = posix_socket_read(file, buf->uio_iov->iov_base, buf->uio_iov->iov_len);
+	if (ret < 0) {
+		ret = errno;
+		goto EXIT;
+	}
+
+	buf->uio_resid -= ret;
+
```

There is a bug here. The function should return an error code (or 0) and not the bytes read.

Add:
`ret = 0;`